streamz/core.py

```diff
@@ -867,12 +867,16 @@ def __init__(self, upstream, func, start=no_default, returns_state=False,
         self.returns_state = returns_state
         # this is one of a few stream specific kwargs
         stream_name = kwargs.pop('stream_name', None)
+        self.sdf_checkpoint = kwargs.pop('sdf_checkpoint', False)
         Stream.__init__(self, upstream, stream_name=stream_name)
 
     def update(self, x, who=None, metadata=None):
         if self.state is no_default:
             self.state = x
-            return self._emit(x, metadata=metadata)
+            if self.sdf_checkpoint:
+                return self._emit((self.state, x), metadata=metadata)
```

Comment:
I've read through the above comments so maybe I have a better idea of what is going on now.

I agree with @martindurant that `sdf_checkpoint` might be a bit too on the nose as names go, since the output state could be used for any number of reasons beyond checkpointing.

My issue here is that the change of that flag causes the graph topology to be different, causing the pipeline to need to be rewritten. I can't think of any current node types that have this kind of behavior.

I don't have a great work around for this though. 

Maybe we could have it named something else so that it is clear that these two modes aren't interchangeable, but this could cause an explosion in the number of classes we track. 

You could add a callback that is called before every emit that takes the current state, but that completely removes it from the graph preventing it from accessing the event loop integration or any other graph based tooling.

You could write a specialty node that when emitted to by the accumulate node access that node's status and emits that. This could maybe be done with a rather odd looking map.
```python
s = Stream()
a = s.accululate(...)
out = a.map(...)...
to_state_store = a.map(lambda x: a.state)
to_state_store.sink(store)
```
As an independent class
```python
class StateGetter(Stream):
    def __init__(self, upstream):
        # XXX: check that upstream has a state if not fail
        self.upstream = upstream
        ...
    def update(self, x):
        return upstream.state
```
This might be one of the more promising approaches, it retains the original topology of the graph, the state storage is on the graph's event loop, and the class version of this even can fail at init time if used improperly.

Reply:
> `sdf_checkpoint` might be a bit too on the nose as names go

Yes, agreed. Will change the name to `with_state`. 

> since the output state could be used for any number of reasons beyond checkpointing.

Yes, that's true. But I haven't seen any streaming stacks to save internal state (which ends up being really complex) for any other reason than checkpointing. Just curious to understand what other cases we might need to save the state for. 

> I don't have a great work around for this though. 

I completely agree. But, as I mentioned above, this PR is targeted primarily for SDF operations to accept an initial state (step 2 in the PR description). @jsmaupin and I will be working on step 3 (publishing the updated SDF at the end of every batch/update). 

This step would handle the tuple of `(state, result)` emitted by the `accumulate` node,  or the `result` of some of the SDF group-by aggs (which is same as the state in these cases). So far, the idea I had to handle this was similar to yours: we add a `publish` node which the SDF aggregation node emits to. The `publish` node plucks the `state` out to the `metadata` already being used for checkpointing, and continues emitting only the `result` downstream. When the checkpointing mechanism is triggered to commit Kafka offsets, we add another callback to checkpoint/save this SDF state either on disk/S3/Dask. 

But, until this gets implemented, I thought we let this be as is, and let the downstream take care of how the state is handled. Hence, the `with_state` parameter is orthogonal to the `start` parameter — one might want to start the job with an SDF state, but might not need the state downstream anymore. But in order to get an initial state (which is usually a complex object like deque, dataframe, etc.), one would need the state downstream some time or another. But again, it's completely discretionary. 

Again, since this is something that is a production-mandated feature; users that don't want to use Kafka as a source, but still want the SDF state downstream, or just want to start  SDFs from a start  state, can handle the `state` however they wish downstream. For us, tying it to the Kafka checkpoint metadata is the eventual goal.

Reply:
I think I agree with all of this. It is OK, from my view, to have a node which outputs different sorts of things depending on an init argument.